modules/core/keeper/msg_server.go

```diff
@@ -70,17 +70,26 @@ func (k Keeper) UpdateClient(goCtx context.Context, msg *clienttypes.MsgUpdateCl
 func (k Keeper) UpgradeClient(goCtx context.Context, msg *clienttypes.MsgUpgradeClient) (*clienttypes.MsgUpgradeClientResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	upgradedClient, err := clienttypes.UnpackClientState(msg.ClientState)
+	// Backwards Compatibility: the light client module is expecting
+	// to unmarshal an interface so we marshal the client state Any
+	upgradedClientState, err := k.cdc.Marshal(msg.ClientState)
```

Comment:
It's a bit unfortunate that we're marshalling here and then unmarshalling later in the callstack. Seems wasteful...

Reply:
I don't think this is necessary. I think this is symptom of how the tests were/are written and they just need to be refactored. After looking closer at the code and how it works, I think it should be possible to pass the `Any.Value` bytes along here and into the light client module and have it work.

Opened a draft which I'll pick up in the morning, but see here for happy path test working: https://github.com/cosmos/ibc-go/pull/5967/files#diff-cef346df04f3fb2455741328504cc2cbb06c7133cce6ede300347139fbdda74eR490

tldr; test code needs to define `*codectypes.Any` and pass their value bytes as args. We can then make the changes in the light client module to unmarshal to concrete types. When we reconstruct the clientState which is proven it has always used `MarshalInterface` given the concrete type and should continue to work like that. I think the tests just need more attention.